utopia-remix/server.js

```diff
@@ -0,0 +1,162 @@
+ \\  \\:\\ /  /:/ /__/:/\\:\\   \\  \\:\\ /  /:/ \\  \\:\\/:/      \\  \\:\\/\\ \\  \\:\\/:/__\\/
+  \\  \\:\\  /:/  \\__\\/  \\:\\   \\  \\:\\  /:/   \\  \\::/        \\__\\::/  \\  \\::/
+   \\  \\:\\/:/        \\  \\:\\   \\  \\:\\/:/     \\  \\:\\        /__/:/    \\  \\:\\
+    \\  \\::/          \\__\\/    \\  \\::/       \\  \\:\\       \\__\\/      \\  \\:\\
+     \\__\\/                     \\__\\/         \\__\\/                   \\__\\/
+`
+
+// -----------------------------------------------------------------------------
+// From: https://github.com/remix-run/remix/tree/main/templates/express
+// -----------------------------------------------------------------------------
+const BUILD_PATH = path.resolve('build/index.js')
+const VERSION_PATH = path.resolve('build/version.txt')
+
+sourceMapSupport.install({
+  retrieveSourceMap: function (source) {
+    const match = source.startsWith('file://')
+    if (match) {
+      const filePath = url.fileURLToPath(source)
+      const sourceMapPath = `${filePath}.map`
+      if (fs.existsSync(sourceMapPath)) {
+        return {
+          url: source,
+          map: fs.readFileSync(sourceMapPath, 'utf8'),
+        }
+      }
+    }
+    return null
+  },
+})
+installGlobals()
+
+async function createDevRequestHandler(initialBuild) {
+  let build = initialBuild
+  async function handleServerUpdate() {
+    // 1. re-import the server build
+    build = await reimportServer()
+    // 2. tell Remix that this app server is now up-to-date and ready
+    broadcastDevReady(build)
+  }
+  const chokidar = await import('chokidar')
+  chokidar
+    .watch(VERSION_PATH, { ignoreInitial: true })
+    .on('add', handleServerUpdate)
+    .on('change', handleServerUpdate)
+
+  // wrap request handler to make sure its recreated with the latest build for every request
+  return async (req, res, next) => {
+    try {
+      return createRequestHandler({
+        build,
+        mode: 'development',
+      })(req, res, next)
+    } catch (error) {
+      next(error)
+    }
+  }
+}
+
+async function reimportServer() {
+  const stat = fs.statSync(BUILD_PATH)
+
+  // convert build path to URL for Windows compatibility with dynamic `import`
+  const BUILD_URL = url.pathToFileURL(BUILD_PATH).href
+
+  // use a timestamp query parameter to bust the import cache
+  return import(BUILD_URL + '?t=' + stat.mtimeMs)
+}
+
+const initialBuild = await reimportServer()
+
+const remixHandler =
+  process.env.NODE_ENV === 'development'
+    ? await createDevRequestHandler(initialBuild)
+    : createRequestHandler({ build: initialBuild })
+// -----------------------------------------------------------------------------
+
+function proxy(originalRequest, originalResponse) {
+  const proxyRequest = http.request(
+    {
+      // target the right server
+      protocol: originalRequest.protocol + ':',
+      host: process.env.EXPRESS_PROXY_TARGET_HOST,
+      port: process.env.EXPRESS_PROXY_TARGET_PORT,
```

Comment:
Side note, but we really need to consolidate a bunch of these environment variables that have been introduced for the remix app. Not as part of this PR though

Reply:
> it'd be great to give this whole thing a cleaning pass

absolutely a side note, but the best time to give it a cleaning pass is _before_ merging the PR :) the second best time is right after